File: tests/internal/fake_engine.cpp
// Test double for the C++ adapter. No device operations or data transfers.
#include "engine.hpp"
#include <algorithm>
#include <utility>

using namespace rdma101;
namespace {
std::string metadata;
void *buffer = nullptr;
std::vector<Request> saved;
uint64_t submissions = 0;
constexpr uint64_t remote = 0xfedcba9876543210;
constexpr uint64_t compare = 0x1122334455667788;
constexpr uint64_t value = 0xffeeddccbbaa0099;
constexpr uint8_t memory_metadata[] = {0, 255, 'M', 'R', 128};
} // namespace

ResultCode Engine::create(Config config, std::unique_ptr<Engine> &engine) {
    metadata = config.device.value_or("<auto>");
    for (auto n : {config.port, config.gid_index}) {
        metadata += "|" + std::to_string(n);
    }
    buffer = nullptr;
    saved.clear();
    submissions = 0;
    engine = std::make_unique<Engine>();
    return ResultCode::Ok;
}
Engine::~Engine() = default;

ResultCode Engine::register_local_memory(void *address,
                                         std::size_t length,
                                         MemoryAccess access,
                                         MemoryMetadata &output) {
    unsigned mask = 0;
    if ((access & MemoryAccess::LocalWrite) != MemoryAccess::None)
        mask |= 1;
    if ((access & MemoryAccess::RemoteWrite) != MemoryAccess::None)
        mask |= 2;
    if ((access & MemoryAccess::RemoteRead) != MemoryAccess::None)
        mask |= 4;
    if ((access & MemoryAccess::RemoteAtomic) != MemoryAccess::None)
        mask |= 8;
    buffer = address;
    output = {77, memory_metadata};
    return length == mask + 1 ? ResultCode::Ok : ResultCode::Invalid;
}
ResultCode Engine::deregister_local_memory(MemoryId memory) {
    return memory == 77 ? ResultCode::Ok : memory == 78 ? ResultCode::Busy : ResultCode::Invalid;
}
ResultCode
Engine::import_remote_memory(PeerId peer, std::span<const uint8_t> bytes, RemoteMemoryId &output) {
    output = 91;
    return peer == 41 && bytes.size() == sizeof(memory_metadata) &&
                   std::equal(bytes.begin(), bytes.end(), memory_metadata)
               ? ResultCode::Ok
               : ResultCode::Invalid;
}
ResultCode Engine::remove_remote_memory(RemoteMemoryId memory) {
    return memory == 91 ? ResultCode::Ok : memory == 92 ? ResultCode::Busy : ResultCode::Invalid;
}
ResultCode Engine::create_peer(PeerMetadata &output) {
    output = {41, {reinterpret_cast<const uint8_t *>(metadata.data()), metadata.size()}};
    return ResultCode::Ok;
}
ResultCode Engine::connect_peer(PeerId peer, std::span<const uint8_t> bytes) {
    return peer == 41 && bytes.size() == metadata.size() &&
                   std::equal(bytes.begin(),
                              bytes.end(),
                              reinterpret_cast<const uint8_t *>(metadata.data()))
               ? ResultCode::Ok
               : ResultCode::Invalid;
}

ResultCode Engine::destroy_peer(PeerId peer) {
    return peer == 41 ? ResultCode::Ok : peer == 42 ? ResultCode::IoError : ResultCode::Invalid;
}

ResultCode Engine::submit_transfer(std::vector<Request> requests, BatchId &batch) {
    ++submissions;
    if (requests.size() != 8)
        return ResultCode::Invalid;
    for (const auto &request : requests) {
        if (request.peer != 41 || request.local_address != buffer || request.length != 8)
            return ResultCode::Invalid;
    }
    const auto *write = std::get_if<Write>(&requests[0].operation);
    const auto *read = std::get_if<Read>(&requests[1].operation);
    const auto *send = std::get_if<Send>(&requests[2].operation);
    const auto *imm = std::get_if<WriteImm>(&requests[4].operation);
    const auto *cas = std::get_if<CompareSwap>(&requests[5].operation);
    const auto *faa = std::get_if<FetchAdd>(&requests[6].operation);
    const auto *plain_send = std::get_if<Send>(&requests[7].operation);
    if (!write || write->remote_address != remote || !write->inline_data || !read ||
        read->remote_address != remote || !send || !send->inline_data ||
        !std::holds_alternative<Recv>(requests[3].operation) || !imm ||
        imm->remote_address != remote || imm->immediate != 0x89abcdef || !imm->inline_data ||
        !cas || cas->remote_address != remote || cas->compare != compare || cas->value != value ||
        !faa || faa->remote_address != remote || faa->value != value || !plain_send ||
        plain_send->inline_data)
        return ResultCode::Invalid;
    saved = std::move(requests);
    batch = submissions;
    return ResultCode::Ok;
}

ResultCode Engine::get_transfer_statuses(BatchId batch, std::span<TransferStatus> output) const {
    if (!output.empty())
        output[0] = Completed{};
    if (batch != submissions || output.size() != saved.size() || saved.empty())
        return ResultCode::Invalid;
    const TransferStatus states[] = {
        Pending{},
        Failed{ResultCode::IoError},
        Completed{Message{saved[2].length}},
        Completed{WriteNotice{std::get<WriteImm>(saved[4].operation).immediate}},
        Failed{ResultCode::Canceled},
        Failed{ResultCode::Unsupported},
        Failed{ResultCode::NoMemory},
        Completed{},
    };
    std::copy(std::begin(states), std::end(states), output.begin());
    return ResultCode::Ok;
}
ResultCode Engine::free_batch(BatchId batch) {
    const ResultCode errors[] = {ResultCode::Unimplemented,
                                 ResultCode::Invalid,
                                 ResultCode::Busy,
                                 ResultCode::IoError,
                                 ResultCode::Timeout,
                                 ResultCode::Unsupported,
                                 ResultCode::NoMemory,
                                 ResultCode::InvalidState,
                                 ResultCode::Canceled};
    return batch <= std::size(errors) ? errors[batch - 1] : ResultCode::Ok;
}

File: tests/internal/fake_engine.rs
// Test double for the Rust adapter. No device operations or data transfers.
pub use crate::native::*;
use std::ffi::c_void;

pub struct Engine {
    metadata: Vec<u8>,
    buffer: *mut c_void,
    saved: Vec<Request>,
    submissions: u64,
}

impl Engine {
    pub fn create(config: Config) -> Result<Box<Self>, Error> {
        let mut metadata = config.device.unwrap_or_else(|| "<auto>".into());
        for n in [config.port, config.gid_index] {
            metadata.push_str(&format!("|{n}"));
        }
        Ok(Box::new(Self {
            metadata: metadata.into_bytes(),
            buffer: std::ptr::null_mut(),
            saved: Vec::new(),
            submissions: 0,
        }))
    }
    pub unsafe fn register_local_memory(
        &mut self,
        address: *mut c_void,
        length: usize,
        access: MemoryAccess,
    ) -> Result<MemoryMetadata<'_>, Error> {
        let mut mask = 0;
        for (flag, bit) in [
            (MemoryAccess::LOCAL_WRITE, 1),
            (MemoryAccess::REMOTE_WRITE, 2),
            (MemoryAccess::REMOTE_READ, 4),
            (MemoryAccess::REMOTE_ATOMIC, 8),
        ] {
            if access.contains(flag) {
                mask |= bit;
            }
        }
        self.buffer = address;
        if length == mask + 1 {
            Ok(MemoryMetadata {
                memory: 77,
                metadata: &[0, 255, b'M', b'R', 128],
            })
        } else {
            Err(Error::Invalid)
        }
    }
    pub unsafe fn deregister_local_memory(&mut self, memory: MemoryId) -> Result<(), Error> {
        match memory {
            77 => Ok(()),
            78 => Err(Error::Busy),
            _ => Err(Error::Invalid),
        }
    }
    pub fn import_remote_memory(
        &mut self,
        peer: PeerId,
        metadata: &[u8],
    ) -> Result<RemoteMemoryId, Error> {
        if peer == 41 && metadata == [0, 255, b'M', b'R', 128] {
            Ok(91)
        } else {
            Err(Error::Invalid)
        }
    }
    pub fn remove_remote_memory(&mut self, memory: RemoteMemoryId) -> Result<(), Error> {
        match memory {
            91 => Ok(()),
            92 => Err(Error::Busy),
            _ => Err(Error::Invalid),
        }
    }
    pub fn create_peer(&mut self) -> Result<PeerMetadata<'_>, Error> {
        Ok(PeerMetadata {
            peer: 41,
            metadata: &self.metadata,
        })
    }
    pub fn connect_peer(&mut self, peer: PeerId, metadata: &[u8]) -> Result<(), Error> {
        if peer == 41 && metadata == self.metadata {
            Ok(())
        } else {
            Err(Error::Invalid)
        }
    }
    pub fn destroy_peer(&mut self, peer: PeerId) -> Result<(), Error> {
        match peer {
            41 => Ok(()),
            42 => Err(Error::Io),
            _ => Err(Error::Invalid),
        }
    }
    pub unsafe fn submit_transfer(&mut self, requests: Vec<Request>) -> Result<BatchId, Error> {
        self.submissions += 1;
        if requests.len() != 8
            || requests
                .iter()
                .any(|r| r.peer != 41 || r.local_address != self.buffer || r.length != 8)
        {
            return Err(Error::Invalid);
        }
        let remote = 0xfedcba9876543210;
        let compare_value = 0x1122334455667788;
        let operand = 0xffeeddccbbaa0099;
        let valid = matches!(requests[0].operation, Operation::Write { remote_address, inline_data: true } if remote_address == remote)
            && matches!(requests[1].operation, Operation::Read { remote_address } if remote_address == remote)
            && matches!(requests[2].operation, Operation::Send { inline_data: true })
            && matches!(requests[3].operation, Operation::Recv)
            && matches!(requests[4].operation, Operation::WriteImm { remote_address, immediate, inline_data: true }
                if remote_address == remote && immediate == 0x89abcdef)
            && matches!(requests[5].operation, Operation::CompareSwap { remote_address, compare, value }
                if remote_address == remote && compare == compare_value && value == operand)
            && matches!(requests[6].operation, Operation::FetchAdd { remote_address, value }
                if remote_address == remote && value == operand)
            && matches!(
                requests[7].operation,
                Operation::Send { inline_data: false }
            );
        if !valid {
            return Err(Error::Invalid);
        }
        self.saved = requests;
        Ok(self.submissions)
    }
    pub fn get_transfer_statuses(
        &self,
        batch: BatchId,
        output: &mut [TransferStatus],
    ) -> Result<(), Error> {
        if let Some(first) = output.first_mut() {
            *first = TransferStatus::Completed { receive: None };
        }
        if batch != self.submissions || output.len() != self.saved.len() || self.saved.is_empty() {
            return Err(Error::Invalid);
        }
        let Operation::WriteImm { immediate, .. } = self.saved[4].operation else {
            unreachable!()
        };
        output.copy_from_slice(&[
            TransferStatus::Pending,
            TransferStatus::Failed(Error::Io),
            TransferStatus::Completed {
                receive: Some(ReceiveResult::Message {
                    length: self.saved[2].length,
                }),
            },
            TransferStatus::Completed {
                receive: Some(ReceiveResult::WriteNotice { immediate }),
            },
            TransferStatus::Failed(Error::Canceled),
            TransferStatus::Failed(Error::Unsupported),
            TransferStatus::Failed(Error::NoMemory),
            TransferStatus::Completed { receive: None },
        ]);
        Ok(())
    }
    pub fn free_batch(&mut self, batch: BatchId) -> Result<(), Error> {
        let errors = [
            Error::Unimplemented,
            Error::Invalid,
            Error::Busy,
            Error::Io,
            Error::Timeout,
            Error::Unsupported,
            Error::NoMemory,
            Error::InvalidState,
            Error::Canceled,
        ];
        if batch <= errors.len() as u64 {
            Err(errors[batch as usize - 1])
        } else {
            Ok(())
        }
    }
}
